Beta 1 draft (#40)
githedgehog/docs: docs/install-upgrade/build-wiring.md

!!! warning ""
    Under construction.

> You can find more details in the User Guide including [switch features and port naming](../user-guide/profiles.md). It's mandatoy to for all switches to reference a `SwitchProfile` in the `spec.profile` of the `Switch` object. Only port naming defined by switch profiles could be used in the wiring diagram, NOS (or any other) port names aren't supported.

Author: You bring up the unwrapping, it is a preference I didn't know that I had until I started doing this and you pointed it out. So a "found" preference. I find it easier when doing mass changes to text that is prose.

In the meantime, to have a look at working wiring diagram for Hedgehog Fabric, run the sample generator that produces
VLAB-compatible wiring diagrams:

```console
ubuntu@sl-dev:~$ hhfab sample -h

NAME:
   hhfab sample - generate sample wiring diagram

USAGE:
   hhfab sample command [command options]

COMMANDS:
   spine-leaf, sl      generate sample spine-leaf wiring diagram
   collapsed-core, cc  generate sample collapsed-core wiring diagram
   help, h             Shows a list of commands or help for one command

OPTIONS:
   --help, -h  show help
```

## Design Discussion
This section is meant to help the reader understand how to assemble the primitives presented by the Fabric API into a functional fabric. This discussion starts with the primary building block of your design, a VPC.

### VPC

> A VPC allows for isolation at layer 3. This is the main building block for users when creating their architecture. Hosts inside of a VPC will see each other but nothing else. The hosts inside of a VPC will likely need to connect to other VPCs or the outside world. To communicate between two VPC a *peering* will need to be created. A VPC can be a logical separation of workloads. By separating these workloads additional controls are available. The logical separation doesn't have to be the traditional database, web, and compute layers it could be development teams who need isolation, it could tenants inside of an office building, or any separation that allows for better control of the network. Once your VPCs are decided, the rest of the fabric will come together. With the VPCs decided traffic can be prioritized, security can be put into place, and the wiring can begin. The fabric allows for the VPC to exist beyond a single switch, which gives flexibility when the physical world meets the digital.

Reviewer: ```suggestion
A VPC allows for isolation at layer 3. This is the main building block for users when creating their architecture. Hosts inside of a VPC will see each other but nothing else. The hosts inside of a VPC will likely need to connect to other VPCs or to the outside world. To communicate between two VPCs, a *peering* will need to be created. A VPC can be a logical separation of workloads. By separating these workloads, additional controls are available. The logical separation doesn't have to be the traditional database, web, and compute layers; it could be development teams who need isolation, or tenants inside of an office building, or any separation that allows for better control of the network. Once you have decided how to setup your VPCs, the rest of the fabric will come together. You can then prioritize traffic, put security policies into place, and begin the wiring. The fabric allows for the VPC to exist beyond a single switch, which gives flexibility when associating physical hardware to your cloud infrastructure.
```

I'd avoid passive tense and figures of speech as much as possible. I'm not very happy with my suggestion for the end of the paragraph though, there's certainly something better to do.

Note: I've got a personal preference for wrapping the text. Most of the sources are wrapped, as far as I can remember.

Reviewer 2: Hosts inside of a VPC will see each other but nothing else.
should be 
host insides of  vpc belong to the same broadcast domain and can see each other.

> A VPC allows for isolation at layer 3. This is the main building block for users when creating their architecture. Hosts inside of a VPC will see each other but nothing else. The hosts inside of a VPC will likely need to connect to other VPCs or the outside world. To communicate between two VPC a *peering* will need to be created. A VPC can be a logical separation of workloads. By separating these workloads additional controls are available. The logical separation doesn't have to be the traditional database, web, and compute layers it could be development teams who need isolation, it could tenants inside of an office building, or any separation that allows for better control of the network. Once your VPCs are decided, the rest of the fabric will come together. With the VPCs decided traffic can be prioritized, security can be put into place, and the wiring can begin. The fabric allows for the VPC to exist beyond a single switch, which gives flexibility when the physical world meets the digital.

Reviewer 2: Wiring has to be defined before we talk about VPC's ? 
wiring is the static part a vpc has nothing to do with wiring

### Connection

A connection represents the physical wires in your data center. They connect switches to other switches or switches to servers.

#### Server Connections

> A server connection will require server side configuration as the Fabric configuration abilities do not reach into the end hosts. A server connection can be one of: 

Reviewer: ```suggestion
A server connection requires server-side configuration as the Fabric configuration abilities do not reach into the end hosts. A server connection can be one of: 
```

Reviewer 2: A server connection is a connection used to connect servers. The fabric will configure the server-facing port according to the type of the connection. The configuration of the actual server needs to done by the server owner. Server Name in these kind of connections only server as metadata to identify the connection and we don't validate if the server name/hostname actually matches what is specified in this connection.

- *Unbundled* - a single cable going from switch to server

> - *Bundled* - two or more cables going to a single switch, the server needs to configured for this, Fabric handles the switch

Reviewer: ```suggestion
- *Bundled* - two or more cables going to a single switch. The server needs to configured for this. The Fabric handles the switch.
```

What's “this”?

> - *MCLAG* -  two cables going to two different switches, also called dual homing. The switches will need a fabric link between them

Reviewer: ```suggestion
- *MCLAG* -  two cables going to two different switches, also called dual homing. The switches will need a fabric link between them.
```

> - *ESLAG* - two to four cables going to different switches, also called multi-homing. If four links are used there will be four switches connected to a single server with four NIC ports

Reviewer: ```suggestion
- *ESLAG* - two to four cables going to different switches, also called multi-homing. Using four links causes four switches to be connected to a single server with four NIC ports.
```

#### Fabric Connections

> These serve as connection between switches, their beautiful weave comprises the fabric of the network.

Reviewer: ```suggestion
Fabric connections serve as connections between switches, they form the fabric of the network.
```

Can be confusing. I know, I'm a party-breaker :grimacing:

Reviewer 2: poetry :)

### VPC Peering

> This is what is needed for VPCs to talk to each other. There are two varieties local and remote.

Reviewer: ```suggestion
VPCs need VPC Peerings to talk to each other. VPC Peerings come in two varieties: local and remote.
```

#### Local VPC Peering

> When the VPCs that need to communicate are both on the same switch. An example would be if your database and web front end servers are in the same rack and are able to be physically cabled to the same switch.

Reviewer: ```suggestion
When the VPCs that need to communicate are both on the same switch, they use a local VPC Peering. For example, if your database and web front-end servers are in the same rack and are able to be physically cabled to the same switch, the VPC Peering is local.
```

Reviewer 2: When there is no dedicated  border/peering switch available in the fabric we can use local VPC peering. This kind of peering tries to send traffic between the two VPC's on the switch where either of the VPC's has workloads attached.

#### Remote VPC Peering

> When the VPCs that need to communicate are on different switches. An example would be if your storage and compute servers are in opposite ends of the data center and need to be cabled to different switches.
> 
> 

Reviewer: ```suggestion
When the VPCs that need to communicate are on different switches, they use a remote VPC Peering. For example, if your storage and compute servers are in opposite ends of the data center and need to be cabled to different switches, they need a remote VPC Peering.
```

Reviewer 2: Remote Peering is used when you want to dedicate a switch to peering VPC. This is either done on the border leaf or on a switch where either of the VPC's are not present. This kind of peering allows peer traffic between different VPC's at line rate and is only limited by fabric bandwidth. On the other hand remote peering introduces a few additional hops in the traffic and may cause a small increase in latency.

In the meantime, to have a look at working wiring diagram for Hedgehog Fabric, run the sample generator that produces
VLAB-compatible wiring diagrams:

```console
ubuntu@sl-dev:~$ hhfab sample -h

NAME:
   hhfab sample - generate sample wiring diagram

USAGE:
   hhfab sample command [command options]

COMMANDS:
   spine-leaf, sl      generate sample spine-leaf wiring diagram
   collapsed-core, cc  generate sample collapsed-core wiring diagram
   help, h             Shows a list of commands or help for one command

OPTIONS:
   --help, -h  show help
```

### Sample Switch Configuration
```yaml
apiVersion: wiring.githedgehog.com/v1beta1
kind: Switch
metadata:
  name: ds3000-02
spec:
  boot:
    serial: ABC123XYZ
  role: server-leaf
  description: leaf-2
  profile: celestica-ds3000
  portBreakouts:
    E1/1: 4x10G
    E1/2: 4x10G
    E1/17: 4x25G
    E1/18: 4x25G
    E1/32: 4x25G
  redundancy:
    group: mclag-1
    type: mclag
```

## Design Discussion
This section is meant to help the reader understand how to assemble the primitives presented by the Fabric API into a functional fabric.

### VPC

> A VPC allows for isolation at layer 3. This is the main building block for users when creating their architecture. Hosts inside of a VPC belong to the same broadcast domain and can communicate with each other. The hosts inside of a VPC will likely need to connect to other VPCs or the outside world. To communicate between two VPC a *peering* will need to be created. A VPC can be a logical separation of workloads. By separating these workloads additional controls are available. The logical separation doesn't have to be the traditional database, web, and compute layers it could be development teams who need isolation, it could tenants inside of an office building, or any separation that allows for better control of the network. Once your VPCs are decided, the rest of the fabric will come together. With the VPCs decided traffic can be prioritized, security can be put into place, and the wiring can begin. The fabric allows for the VPC to exist beyond a single switch, which gives flexibility when the physical world meets the digital.

Reviewer 3: * Hosts inside of a vpc do not necessarily belong to the same broadcast domain. A vpc can have several subnets.

* "it could tenants" -> "it could be tenants"
* "which gives flexibility when the physical world meets the digital" ....by Logan Shakespeare ;-) . Now seriously, I think here you could say that a VPC can physically span more than one swicth, which provides great flexibility, for instance for workload mobility.

### Connection

> A connection represents the physical wires in your data center. They connect switches to other switches or switches to servers.

Reviewer 3: data-center -> fabric (?)

#### Server Connections

A server connection is a connection used to connect servers to the fabric. The fabric will configure the server-facing port according to the type of the connection (MLAG, Bundle, etc).The configuration of the actual server needs to be done by the server administrator. The server name is not validated by the fabric and is used as metadata to identify the connection. A server connection can be one of: 

- *Unbundled* - A single cable connecting switch to server.
- *Bundled* - Two or more cables going to a single switch, a LAG or similar.
- *MCLAG* -  Two cables going to two different switches, also called dual homing. The switches will need a fabric link between them.
- *ESLAG* - Two to four cables going to different switches, also called multi-homing. If four links are used there will need to be four switches connected to a single server with four NIC ports.

#### Fabric Connections

Fabric connections serve as connections between switches, they form the fabric of the network.

### VPC Peering

VPCs need VPC Peerings to talk to each other. VPC Peerings come in two varieties: local and remote.

#### Local VPC Peering

> When there is no dedicated border/peering switch available in the fabric we can use local VPC peering. This kind of peering tries to send traffic between the two VPC's on the switch where either of the VPC's has workloads attached. Due to limitation in the Sonic network operating system this kind of peering bandwidth is limited to the number of VPC loopbacks you have selected while initializing the fabric.

Reviewer 3: ", in  this kind of peering, bandwidth is limited ..." ? If I understand the sentence...

#### Remote VPC Peering

Remote Peering is used when you need a high bandwidth connection between the VPCs, you will dedicate a switch to the peering traffic. This is either done on the border leaf or on a switch where either of the VPC's are not present. This kind of peering allows peer traffic between different VPC's at line rate and is only limited by fabric bandwidth. Remote peering introduces a few additional hops in the traffic and may cause a small increase in latency.

#### VPC Loopback

A VPC loopback is a physical cable with both ends plugged into the same switch, suggested but  not required to be the adjacent ports. This loopback allows two different VPCs to communicate with each other.

> 

Reviewer 4: or "limitation"
